Source/WebCore/editing/SpellChecker.cpp

```diff
@@ -210,12 +228,44 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
     m_requestQueue.append(WTFMove(request));
 }
 
-void SpellChecker::didCheck(TextCheckingRequestIdentifier identifier, const Vector<TextCheckingResult>& results)
+static bool containsGrammarResult(TextCheckingResult result, const Vector<TextCheckingResult>& existingResults)
 {
-    ASSERT(m_processingRequest);
-    ASSERT(m_processingRequest->data().identifier() == identifier);
-    if (m_processingRequest->data().identifier() != identifier) {
-        m_requestQueue.clear();
+    bool foundIt = false;
+    for (TextCheckingResult existingResult : existingResults) {
+        TextCheckingType existingResultType = *existingResult.type.toSingleValue();
+        if (existingResultType != TextCheckingType::Grammar || result.range.location != existingResult.range.location || result.range.length != existingResult.range.length || result.details.size() != existingResult.details.size())
+            continue;
+        bool detailsMatch = true;
+        for (auto [detail, existingResultDetail] : zippedRange(result.details, existingResult.details)) {
+            if (!detailsMatch)
+                break;
+            detailsMatch = std::tie(detail.range.location, detail.range.length, detail.guesses) == std::tie(existingResultDetail.range.location, existingResultDetail.range.length, existingResultDetail.guesses);
+        }
```

Comment:
totally optional, but if you would like to adopt it, then `ranges::all_of` is an idiom that expresses this pattern (even the early return!)

```suggestion
        bool detailsMatch = std::ranges::all_of(zippedRange(result.details, existingResult.details), [](const auto& detailPair) {
            auto [detail, existingResultDetail] = detailPair;
            return std::tie(detail.range.location, detail.range.length, detail.guesses) == std::tie(existingResultDetail.range.location, existingResultDetail.range.length, existingResultDetail.guesses);
        });
```

Reply:
no guarantees that it works with `WTF::ZipperedRange`, but here's an example with the standard library analogue https://godbolt.org/z/rad5WMGch